### PR TITLE
Success and Error bars

### LINKED ISSUE #75: Errors on the Info bar
# Goal

Right now, there is some error handling that involves an `eprintln!` and then nothing else. In these cases, rather than informing the user in the terminal, we should inform them using our info bar. 

# Implementation

* Look through all the places where we `eprintln!` on error, and instead send that text to the info bar.
* Look at the command mode errors as an example.
* Use error mode. 
* Color the info bar red instead of blue when there is an error that needs to be displayed (also for existing error info in command mode)

### PR DESCRIPTION
**Pull request by Davejkane**
This PR closes [#75](https://github.com/Davejkane/riv/issues/75) and [#78](https://github.com/Davejkane/riv/issues/78) and

### CODE REVIEW

File: src/program/mod.rs [hunk 1]
```diff
@@ -346,12 +346,16 @@ impl<'a> Program<'a> {
             ))
         });
         let newname = self.construct_dest_filepath(filepath)?;
+        let cloned_name = newname.clone();
```
Comment (gurgalex) [inline src/program/mod.rs]: Clone can be eliminated by passing `newname` as a reference to `copy`

File: src/program/mod.rs [hunk 2]
```diff
@@ -364,6 +368,7 @@ impl<'a> Program<'a> {
                 self.paths.index, self.paths.max_viewable
             ))
         });
+        let cloned_path = current_imagepath.clone();
```
Comment (gurgalex) [inline src/program/mod.rs]: Same to eliminate the clone here as well. (&current_imagepath for `move`)

File: src/program/mod.rs [hunk 3]
```diff
@@ -405,6 +415,7 @@ impl<'a> Program<'a> {
...
```
Comment (gurgalex) [inline src/program/mod.rs]: ~Clone elimination as well.~
Update: Reference may not exist anymore, so it makes sense to clone the PathBuf.

File: src/program/mod.rs [hunk 4]
```diff
@@ -496,16 +515,35 @@ impl<'a> Program<'a> {
...
                     Action::Copy => match self.copy_image() {
-                        Ok(_) => (),
-                        Err(e) => eprintln!("Failed to copy file: {}", e),
+                        Ok(s) => {
+                            self.ui_state.mode = Mode::Success(s);
+                            return Ok(());
+                        }
+                        Err(e) => {
+                            self.ui_state.mode = Mode::Error(format!("Failed to copy file: {}", e));
+                            return Ok(());
```
Comment (gurgalex) [inline src/program/mod.rs:523]: So copying was `Ok` when it failed?

I guess we need internal error handling (enums / failure) to improve the error flow.

File: src/program/render.rs [hunk 5]
```diff
@@ -213,24 +214,25 @@ impl<'a> Program<'a> {
...
 
     fn render_bar(&mut self, dims: (u32, u32, u32)) -> Result<(), String> {
+        let colors = mode_colors(&self.ui_state.mode);
         let height = dims.0;
         let width = self.screen.canvas.viewport().width();
         let y = (self.screen.canvas.viewport().height() - height) as i32;
         let mut x = 0;
         let mut w = dims.1 + HALF_PAD as u32 * 3;
-        self.screen.canvas.set_draw_color(light_blue());
+        self.screen.canvas.set_draw_color(colors.0);
```
Comment (gurgalex) [inline src/program/render.rs]: I'm not sure what `.0` `.1` or `.2` mean.

File: src/program/render.rs [hunk 6]
```diff
@@ -263,6 +265,16 @@ impl<'a> Program<'a> {
     }
 }
 
+fn mode_colors(m: &Mode) -> (Color, Color, Color) {
```
Comment (gurgalex) [inline src/program/render.rs]: Maybe explain the three fields here, and/or make it return a struct with named fields.

File: src/program/render.rs [hunk 7]
```diff
@@ -263,6 +265,16 @@ impl<'a> Program<'a> {
...
+    match m {
+        Mode::Normal => (light_blue(), blue(), grey()),
+        Mode::Error(_) => (light_red(), red(), grey()),
```
Comment (gurgalex) [inline src/program/render.rs]: The Black text and darkish red background are somewhat too low contrast.

Maybe a slightly lighter red with white text instead.

**Review by gurgalex — CHANGES_REQUESTED** [review]
@Davejkane Thanks for making all the errors render to the infobar.

I'm getting the feeling that proper internal error handling is really needed (but that can be done in a later PR).
https://rust-lang-nursery.github.io/failure/

Changes

A description (or struct) about what each field in the tuple returned from `make_colors` does.

A Higher contrast error bar (red bar with white text perhaps)

File: src/program/mod.rs (hunk 2 above)
Comment (gurgalex) [reply src/program/mod.rs]: Hmm, Borrow checker, I see. The reference to the path may not exist after it has been removed.

**Comment by Davejkane** [comment]
@gurgalex thanks for the review. I think I've handled the changes, please have another look.

**Comment by gurgalex** [comment]
Much better. Thank you.

**Review by gurgalex — APPROVED** [review]

**Comment by gurgalex** [comment]
Hmm, found something.

When a new glob is successfully done, the success message hides the image path the first image until a button is pressed.

**Pull request merged**